### PR TITLE
👷🏼‍♀️ Add IAC and deploy pipeline 🚧

### PR DESCRIPTION
**Pull request by lisamittmann**
- add terraform code for iac set up 
- add deployment pipeline with build frontend, build backend, build image with jib and deploy to cloud run
- fix some frontend and backend issues to enable frontend and backend deploy

### CODE REVIEW

File: frontend/src/App.js [hunk 1]
```diff
@@ -1,13 +1,18 @@
...
-            <RecipePage/>
-        </Route>
+        <div>
+            <Route exact path="/">
+                <p>This is my starter page</p>
+                <Link to="/recipe/day33">Schokokuchen</Link>
+            </Route>
+            <Route path="/recipe/:recipeId">
+                <RecipePage/>
+            </Route>
```
Comment (christophersiem) [inline frontend/src/App.js]: Have you omitted the Switch on purpose?

**Review by christophersiem — APPROVED** [review]
Can't really comment on .tf files, everything else looked very good 🎉

File: frontend/src/App.js (hunk 1 above)
Comment (lisamittmann) [reply frontend/src/App.js]: very good question 🤣  I simply forgot 🙈

File: .github/workflows/deployment.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Set up Terraform
+        uses: hashicorp/setup-terraform@v1
+
+      - name: Set Up Credentials
+        env:
+          SET_UP_CREDENTIALS: ${{ secrets.GOOGLE_CREDENTIALS }}
+        run: printf '%s' "$SET_UP_CREDENTIALS" > "$GOOGLE_APPLICATION_CREDENTIALS"
```
Comment (fabianschmauder) [inline .github/workflows/deployment.yml:30]:
```suggestion
    env:
      GOOGLE_APPLICATION_CREDENTIALS_PATH: /tmp/credentials.json

    defaults:
      run:
        shell: bash

    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Set up Terraform
        uses: hashicorp/setup-terraform@v1

      - name: Set Up Credentials
        env:
          SET_UP_CREDENTIALS: ${{ secrets.GOOGLE_CREDENTIALS }}
        run: printf '%s' "$SET_UP_CREDENTIALS" > "$GOOGLE_APPLICATION_CREDENTIALS_PATH"
```
i would call it path or directly use it

```suggestion
    defaults:
      run:
        shell: bash

    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Set up Terraform
        uses: hashicorp/setup-terraform@v1

      - name: Set Up Credentials
        env:
          SET_UP_CREDENTIALS: ${{ secrets.GOOGLE_CREDENTIALS }}
        run: printf '%s' "$SET_UP_CREDENTIALS" > "/tmp/credentials.json"
```

File: .github/workflows/deployment.yml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+      - name: Terraform Init
+        run: |
+          cd iac/terraform
+          terraform init
+
+      - name: Terraform Plan
+        run: |
+          cd iac/terraform
+          printf "mongodburi = \"${{ secrets.MONGODB_URI }}\"" > terraform.tfvars
```
Comment (fabianschmauder) [inline .github/workflows/deployment.yml]: Here you use secrets directly at line  29 you first create an env variable:
`SET_UP_CREDENTIALS: ${{ secrets.GOOGLE_CREDENTIALS }}`

I would do it the same way

File: .github/workflows/deployment.yml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+          terraform plan
+
+      - name: Terraform Apply
+        if: github.event_name == 'push'
```
Comment (fabianschmauder) [inline .github/workflows/deployment.yml]: is there a possibility that github.event_name == 'push' is not true?
your job only runs on push
```on:
  push:
    branches:
      - main
   ```

File: iac/terraform/main.tf [hunk 5]
```diff
@@ -0,0 +1,9 @@
+
+module "storage" {
+  source = ".//storage"
+}
+
+module "container" {
+  source = ".//container"
```
Comment (fabianschmauder) [inline iac/terraform/main.tf]:
```suggestion
  source = "./container"
```
is there a reason for .//

**Review by fabianschmauder — COMMENTED** [review]
LTGM 🎉

File: .github/workflows/deployment.yml (hunk 2 above)
Comment (lisamittmann) [reply .github/workflows/deployment.yml:30]: totally get your point but this is actually a requirement from TF plus gcloud, TF expects to find a path in the env var `GOOGLE_APPLICATION_CREDENTIALS` under this very specific name. Therefore, for the TF steps to work with the env var I need to do it like this 😔

File: .github/workflows/deployment.yml (hunk 4 above)
Comment (lisamittmann) [reply .github/workflows/deployment.yml]: maybe 🙈 
removed it

**Pull request merged**